packages/core-flows/src/product/steps/batch-link-products-in-category.ts

```diff
@@ -0,0 +1,91 @@
+import { ModuleRegistrationName } from "@medusajs/modules-sdk"
+import { IProductModuleService, ProductCategoryWorkflow } from "@medusajs/types"
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+export const batchLinkProductsToCategoryStepId =
+  "batch-link-products-to-category"
+export const batchLinkProductsToCategoryStep = createStep(
+  batchLinkProductsToCategoryStepId,
+  async (
+    data: ProductCategoryWorkflow.BatchUpdateProductsOnCategoryWorkflowInput,
+    { container }
+  ) => {
+    const service = container.resolve<IProductModuleService>(
+      ModuleRegistrationName.PRODUCT
+    )
+
+    if (!data.add?.length && !data.remove?.length) {
+      return new StepResponse(void 0, null)
+    }
+
+    const toRemoveSet = new Set(data.remove?.map((id) => id))
+    const dbProducts = await service.list(
+      { id: [...(data.add ?? []), ...(data.remove ?? [])] },
+      {
+        take: null,
+        select: ["id", "categories"],
+      }
+    )
+
+    const productsWithUpdatedCategories = dbProducts.map((p) => {
+      if (toRemoveSet.has(p.id)) {
+        return {
+          id: p.id,
+          category_ids: (p.categories ?? [])
+            .filter((c) => c.id !== data.id)
+            .map((c) => c.id),
```

Comment:
q: should we just pass categories and skip the song and dance between module and workflows?

```suggestion
          categories: (p.categories ?? [])
            .filter((c) => c.id !== data.id)
```

Reply:
Could make sense. We can even go as far as typing the `categories` input as `categories: { id: string }[]`, and then we wouldn't have to normalize anything at either level.

Reply:
I think this is how we do for some other modules as well

Reply:
Keeping it as is for now and creating a ticket